cmd/kubeadm/app/cmd/upgrade/diff.go

```diff
@@ -0,0 +1,141 @@
+	configutil "k8s.io/kubernetes/cmd/kubeadm/app/util/config"
+	"k8s.io/kubernetes/pkg/util/version"
+)
+
+type diffFlags struct {
+	apiServerManifestPath         string
+	controllerManagerManifestPath string
+	schedulerManifestPath         string
+	newK8sVersionStr              string
+	contextLines                  int
+	parent                        *cmdUpgradeFlags
+}
+
+var (
+	defaultAPIServerManifestPath         = constants.GetStaticPodFilepath(constants.KubeAPIServer, constants.GetStaticPodDirectory())
+	defaultControllerManagerManifestPath = constants.GetStaticPodFilepath(constants.KubeControllerManager, constants.GetStaticPodDirectory())
+	defaultSchedulerManifestPath         = constants.GetStaticPodFilepath(constants.KubeScheduler, constants.GetStaticPodDirectory())
+)
+
+// NewCmdDiff returns the cobra command for `kubeadm upgrade diff`
+func NewCmdDiff(parentflags *cmdUpgradeFlags) *cobra.Command {
+	flags := &diffFlags{
+		parent: parentflags,
+	}
+
+	cmd := &cobra.Command{
+		Use:   "diff [version]",
+		Short: "Show what differences would be applied to existing static pod manifests",
+		Run: func(cmd *cobra.Command, args []string) {
+			runDiff(flags, args)
+		},
+	}
+
+	cmd.Flags().StringVar(&flags.apiServerManifestPath, "api-server-manifest", defaultAPIServerManifestPath, "path to API server manifest")
+	cmd.Flags().StringVar(&flags.controllerManagerManifestPath, "controller-manager-manifest", defaultControllerManagerManifestPath, "path to controller manifest")
+	cmd.Flags().StringVar(&flags.schedulerManifestPath, "scheduler-manifest", defaultSchedulerManifestPath, "path to scheduler manifest")
+	cmd.Flags().IntVarP(&flags.contextLines, "context-lines", "c", 3, "How many lines of context in the diff")
+
+	return cmd
+}
+
+func runDiff(flags *diffFlags, args []string) {
+	var cfg *kubeadmapi.MasterConfiguration
+
+	// If the version is specified in config file, pick up that value.
+	if flags.parent.cfgPath != "" {
+		var err error
+		glog.V(1).Infof("fetching configuration from file", flags.parent.cfgPath)
+		cfg, err = configutil.ConfigFileAndDefaultsToInternalConfig(flags.parent.cfgPath, &kubeadmapiext.MasterConfiguration{})
+		kubeadmutil.CheckErr(err)
+
+		if cfg.KubernetesVersion != "" {
+			flags.newK8sVersionStr = cfg.KubernetesVersion
+		}
+	}
+
+	// If the new version is already specified in config file, version arg is optional.
+	if flags.newK8sVersionStr == "" {
+		err := cmdutil.ValidateExactArgNumber(args, []string{"version"})
+		kubeadmutil.CheckErr(err)
+	}
+
+	// If option was specified in both args and config file, args will overwrite the config file.
+	if len(args) == 1 {
+		flags.newK8sVersionStr = args[0]
+	}
+
+	// Convert to an internal representation
+	internalcfg := &kubeadmapi.MasterConfiguration{}
+	scheme.Scheme.Convert(cfg, internalcfg, nil)
+
+	k8sVer, err := version.ParseSemantic(flags.newK8sVersionStr)
+	if err != nil {
+		kubeadmutil.CheckErr(err)
```

Comment:
it's really hard to write unit tests around `CheckErr()` calls as it's effectively calls `exit(1)` on error. :(

Reply:
Can you please file an issue to improve that? Such unit testing would be a perfect contribution from you :+1:

Reply:
possible tests:
- non existing config path in `flags.parent.cfgPath`
- `flags.newK8sVersionStr == ""`, but no `version` arg is provided
- bad version format in `flags.newK8sVersionStr` to make ParseSemantic() fail
- make `kubeadmutil.MarshalToYaml` fail somehow.
- fail `ioutil.ReadFile(path)` by providing bad 'path'

i don't think that the diff output itself should be tested - e.g. making sure that a diff buffer looks exactly like another diff buffer.

one problem i have with Lucas' request is that my setup is quite broken at the moment, so if you want me to add tests for this, i would need a fully working master configuration yaml aaand i might ask some questions too - because i simply don't understand some of the things here.

Reply:
@liztio TLDR if you are out of time and want me to help send me a YAML that returns `nil` for `runDiff()`.

Reply:
Those sound good. Unfortunately, I'm out thursday and friday this week, and it's basically the end of my day. So given the choice between adding unit tests and missing the 0.11 deadline, I'll defer the unit tests until I get back. I [made an issue](https://github.com/kubernetes/kubeadm/issues/826) so I don't forget to follow up.